### DISCUSSION TITLE
UNSAFE_METHOD_ACCESS warning should take into consideration existing "if" statements

### DISCUSSION THREAD

**Discussion by Blocker226** [godotengine/godot-proposals#8112] [Scripting]
Hi, I am a junior developer with previous experience in Unity, transitioning to Godot and GDScript. I used to enjoy JetBrains Rider's C# support, which included auto-formatting for safe casting/method access.

Coming to GDScript, I noticed that when I set the `warnings/unsafe_method_access` setting in the project settings, it does encourage me to use the `as` keyword to cast my derived classes accordingly, providing some level of validation on par with Rider.

However, it does not take into account the presence of any type of checks that precede it, namely, cases like this, where the class `Selectable` extends `Node3D`:
```gdscript
if current_hover && current_hover is Selectable:
    current_hover.set_highlight(true) 
# The method "set_highlight()" is not present on the inferred type "Node3D" (but may be present on a subtype).
```

I'm hoping it's possible to add some sort of contextualisation to the editor for it to be able to identify such cases, and perhaps open up new opportunities to add auto-suggest/format/fix options down the line.

**Comment by AThousandShips** [comment]
It does, as far as I know, either this is a bug or a limitation with using custom classes

Does it work if you do:
```gdscript
if current_hover is Selectable:
    current_hover.set_highlight(true)
```
AFAIK `is` does null checking too 

CC @dalexeev

**Reply by dalexeev** [reply]
The editor takes into account assignments and type checks using the `is` operator when possible, but warnings (and the compiler) only take into account the declared type. Use a new variable in one of the following ways:

```gdscript
var node_2d := node as Node2D
if node_2d:
    # code
```

```gdscript
if node is Node2D:
    var node_2d: Node2D = node
    # code
```

**Reply by Blocker226** [reply]
I can confirm that `is` does null checking, but that code snippet doesn't work either, probably for the reasons @dalexeev mentioned. His workaround does work, but I think this would be nice to look into down the line, at least for style reasons.

Maybe expanding the warning to be able to elaborate on, show or link to workarounds such as these would suffice in the short term?

**Reply by miv391** [reply]
New variable is not necessarily needed. This should also work:

```gdscript
if current_hover and current_hover is Selectable:
	(current_hover as Selectable).set_highlight(true)
```	

But it would be nice if GDScript had an optional chaining operator `?.` like in Typescript. In this example you could write:

```gdscript
(current_hover as Selectable)?.set_highlight(true)
```

`set_highlight` would be called only if `current_hover as Selectable` is not null.

**Reply by Blocker226** [reply]
@miv391 Yes, but I'm trying to avoid repetitive scenarios like this:
```gdscript
if current_hover and current_hover is Selectable:
    (current_hover as Selectable).set_highlight(true)
    (current_hover as Selectable).on_select()
    # code that relies on Selectable
```

I do agree that having a conditional access operator would be nice, but that's probably for another discussion.

**Reply by dalexeev** [reply]
Yes, it is not necessary to create a new variable. But keep in mind that the `as` operator in this case generates runtime code, so it is more rational to create a variable if you want to use the result several times.

**Reply by Mickeon** [reply]
Admittedly, it would be nice if the compiler was smarter about this and casted under the hood. I aim to have as many safe lines as possible, but to achieve that an implicit cast is always necessary, which feels redundant for GDScript, specifically.

**Comment by mieldepoche** [comment]
see also https://github.com/godotengine/godot-proposals/issues/3691
